placement/cluster_mapping_provider.go

```diff
@@ -0,0 +1,476 @@
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+import (
+	"errors"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3storage"
+	"github.com/m3db/m3storage/generated/proto/schema"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/time"
+
+	"github.com/facebookgo/clock"
+	"github.com/willf/bitset"
+)
+
+var (
+	errInvalidTransition = errors.New("invalid shard transition")
+)
+
+// ClusterMappingProviderOptions are options used to build a ClusterMappingProvider
+type ClusterMappingProviderOptions interface {
+	// Logger is the logger to use
+	Logger(logger xlog.Logger) ClusterMappingProviderOptions
+	GetLogger() xlog.Logger
+
+	// Clock is the clock to use
+	Clock(clock clock.Clock) ClusterMappingProviderOptions
+	GetClock() clock.Clock
+}
+
+// NewClusterMappingProviderOptions returns new empty ClusterMappingProviderOptions
+func NewClusterMappingProviderOptions() ClusterMappingProviderOptions {
+	return clusterMappingProviderOptions{}
+}
+
+// clusterMappingProvider implements the ClusterMappingProvider interface on
+// top of the placement data
+type clusterMappingProvider struct {
+	sync.RWMutex
+	byRetention []*databaseMappings          // ordered by retention period (shortest first)
+	byName      map[string]*databaseMappings // indexed by name
+
+	watch  kv.ValueWatch
+	closed chan struct{}
+	log    xlog.Logger
+	clock  clock.Clock
+}
+
+// NewClusterMappingProvider returns a new ClusterMappingProvider
+// around a kv store and key
+func NewClusterMappingProvider(
+	key string,
+	kvStore kv.Store,
+	opts ClusterMappingProviderOptions,
+) (storage.ClusterMappingProvider, error) {
+	if opts == nil {
+		opts = clusterMappingProviderOptions{}
```

Comment:
Any reason not to call `NewClusterMappingProviderOptions()` to centralize options construction?